File: leetcode/20.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) == 1:
            return False
        stack = []
        brackets = {')': '(', ']': '[', '}': '{'}
        
        for i in range(len(s)):
            if s[i] == '(' or s[i] == '[' or s[i] == '{':
                stack.append(s[i])
            if s[i] == ')' or s[i] == ']' or s[i] == '}':
                if stack == []:
                    return False
                elif stack[-1] == brackets[s[i]]:
                    stack.pop()
                else:
                    return False
       
        if stack == []:
            return True
        else:
            return False

File: code/20.java
class Solution {
    public boolean isValid(String s) {
        char[] stack = new char[s.length()];

        int index = -1;

        for (char c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                stack[++index] = c;
            } else {
                if (index == -1) return false;
                else if ((c == ')' && stack[index] == '(') ||
                    (c == '}' && stack[index] == '{') ||
                    (c == ']' && stack[index] == '[')) index--;
                else return false;
            }
        }

        return index == -1;
    }
}